### PR TITLE
Add TCLAP article

### PR DESCRIPTION
**Pull request by ccombei-esrlabs**
How to use TCLAP to parse the c++ dumpsys command line.

### CODE REVIEW

File: content/posts/tclap.md [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    }
+
+    dprintf(fd, "%s", buffer.str().c_str());
+    ::std::cerr.rdbuf(old_cerr); // reset to standard error again
+    ::std::cout.rdbuf(old_cout); // reset to standard output again
+    return EX_NONE;
+}
+```
+
+## Constraints
```
Comment (elektronenhirn) [inline content/posts/tclap.md:65]: Consider "Android specifics" instead of "Constraints", just because "Constrains" sounds so negative ;-)

**Review by elektronenhirn — CHANGES_REQUESTED** [review]

File: content/posts/tclap.md [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+date: 2023-09-06T17:49:38+02:00
+draft: false
+author: "The Android Team"
+---
+
+## Introduction
+The `Templatized C++ Command Line Parser` is a c++ header library that is used to parse command line arguments. The developers of this library maintain a manual (https://tclap.sourceforge.net/manual.html) that is easy to use, so be sure to check it out.
+
+## Usage
+In order to use this library to parse your dumpsys commands, you can just copy the following code and adapt the todos to fit your needs
```
Comment (cipt2001) [inline content/posts/tclap.md:12]: Any includes needed? Any libraries to add to Android.bp?

File: content/posts/tclap.md [hunk 3]
```diff
@@ -0,0 +1,74 @@
+---
+title: "Using TCLAP to parse dumpsys commands"
+date: 2023-09-06T17:49:38+02:00
+draft: false
+author: "The Android Team"
```
[the remark this answers is not in the document]
Comment (ccombei-esrlabs) [reply content/posts/tclap.md]: done

File: content/posts/tclap.md [hunk 4]
```diff
@@ -0,0 +1,74 @@
+---
+title: "Using TCLAP to parse dumpsys commands"
+date: 2023-09-06T17:49:38+02:00
```
[the remark this answers is not in the document]
Comment (ccombei-esrlabs) [reply content/posts/tclap.md]: done

File: content/posts/tclap.md [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+* **Redirect the output of cout and cerr to a string stream:** because dumpsys uses a custom file descriptor and TCLAP uses std::cout and std::cerr function calls, it is necessary that we redirect the std::cout and std::cerr outputs to a local buffer that we will print to the custom file descriptor. This approach creates one constraint: the function must have only one return point. If you want to have multiple return points, then before each return statement, you must print to the custom file descriptor and reset the standard output and error again.
+* **Handle all exceptions:** by default, TCLAP handles the exceptions on its own. If an exception is detected, then TCLAP outputs a log and then calls the "exit" function, which will result in the operating system killing the process. This is why we must tell TCLAP to not handle any exceptions and to simply throw them. If we do not catch these exceptions, then they will be thrown all the way to lib bionic, which will also
```
[the remark this answers is not in the document]
Comment (ccombei-esrlabs) [reply content/posts/tclap.md]: done

File: content/posts/tclap.md (hunk 1 above)
Comment (ccombei-esrlabs) [reply content/posts/tclap.md]: done

File: content/posts/tclap.md (hunk 2 above)
Comment (ccombei-esrlabs) [reply content/posts/tclap.md:28]: done

**Review by cipt2001 — APPROVED** [review]

**Review by elektronenhirn — APPROVED** [review]

**Pull request merged**